### PR TITLE
Infra docker with dummy frontend

### PR DESCRIPTION
**Pull request by JaafarRammal**
# Summary

Docker encapsulation with server in go. Added flag to trigger server, defaults to normal backend simulation run behaviour.

### CODE REVIEW

**Review by udai-arneja — DISMISSED** [review]
LGTM, this was fun

File: .dockerignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+**/node_modules
+**/build
+**/logs
+**/.git
```
Comment (jzzheng22) [inline .dockerignore:4]:
```suggestion
**/.git

```

Comment (jzzheng22) [reply .dockerignore:4]: GitHub complaining about no newline at end of file so might as well change it I guess (or not... I don't really care)

File: .github/workflows/package.yaml [hunk 2]
```diff
@@ -30,8 +33,63 @@ jobs:
...
+          [[ "${{ github.ref }}" == "refs/tags/"* ]] && VERSION=$(echo $VERSION | sed -e 's/^v//')
+
+          # Use Docker `latest` tag convention
+          [ "$VERSION" == "main" ] && VERSION=latest
+
+          echo IMAGE_ID=$IMAGE_ID
+          echo VERSION=$VERSION
+
+          docker tag $IMAGE_NAME $IMAGE_ID:$VERSION
+          docker push $IMAGE_ID:$VERSION
```
Comment (jzzheng22) [inline .github/workflows/package.yaml:95]:
```suggestion
          docker push $IMAGE_ID:$VERSION 
          
```

Comment (jzzheng22) [reply .github/workflows/package.yaml:95]: GitHub complaining about no newline at end of file so might as well change it I guess (or not... I don't really care)

File: .gitignore [hunk 3]
```diff
@@ -21,4 +21,7 @@ bin/
 # Logs
 logs/
 **.log
-simulation.json
\ No newline at end of file
+simulation.json
```
Comment (jzzheng22) [inline .gitignore:24]:
```suggestion
```

Comment (jzzheng22) [reply .gitignore:24]: We don't generate this file anymore so no point having it in gitignore

File: .gitignore [hunk 4]
```diff
@@ -21,4 +21,7 @@ bin/
...
+
+# Frontend
+**/build
```
Comment (jzzheng22) [inline .gitignore:27]: GitHub complaining about no newline at end of file so might as well change it I guess (or not... I don't really care)

File: Dockerfile [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+RUN mkdir /app
+WORKDIR /app
+RUN mkdir /build
+RUN apk add git gcc libc-dev g++
+ADD . /app/
+COPY --from=FRONTEND /app/frontend/build /app/build
+RUN go mod download
+RUN go build -o main main.go
+EXPOSE 9000
+CMD ["/app/main", "--mode=serve"]
```
Comment (jzzheng22) [inline Dockerfile:32]: GitHub complaining about no newline at end of file so might as well change it I guess (or not... I don't really care)
```suggestion
CMD ["/app/main", "--mode=serve"]

```

File: main.go [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+func runNewSimulation(parameters config.ConfigParameters) {
+	rand.Seed(time.Now().UnixNano())
+	f := setupLogFile()
+	healthInfo := health.NewHealthInfo(&parameters)
+
+	// TODO: agentParameters - struct
+	simEnv := simulation.NewSimEnv(&parameters, healthInfo)
+	simEnv.Simulate()
+	f.Close()
```
Comment (jzzheng22) [inline main.go:60]:
```suggestion
	f := setupLogFile()
	defer f.Close()
	healthInfo := health.NewHealthInfo(&parameters)

	// TODO: agentParameters - struct
	simEnv := simulation.NewSimEnv(&parameters, healthInfo)
	simEnv.Simulate()
```

File: main.go [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+}
+
+func setupLogFile() (fp *os.File) {
+	// Setup a log file
```
Comment (jzzheng22) [inline main.go:64]:
```suggestion
```

File: main.go [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+	f := setupLogFile()
+	healthInfo := health.NewHealthInfo(&parameters)
+
+	// TODO: agentParameters - struct
+	simEnv := simulation.NewSimEnv(&parameters, healthInfo)
+	simEnv.Simulate()
+	f.Close()
+}
+
+func setupLogFile() (fp *os.File) {
```
Comment (jzzheng22) [inline main.go:63]: This function should probably return an error as well, and then `runNewSimulation()` checks if there is an error.

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: Makefile [hunk 9]
```diff
@@ -20,13 +20,13 @@ test-coverage: dep ## Run tests with coverage
...
 build: dep ## Build the binary file
-	@go build -o bin/backend ./cmd/backend
+	@go build -o bin/backend .
  
 clean: ## Remove previous build
 	@rm -rf ./bin ./logs
  
 run: ## run simulation
-	@go run cmd/backend/main.go
+	@go run main.go $(ARGS)
```
Comment (jzzheng22) [inline Makefile:29]: Where is `$(ARGS)` defined?

File: Makefile [hunk 10]
```diff
@@ -20,13 +20,13 @@ test-coverage: dep ## Run tests with coverage
...
 
 help: ## Display this help screen
 	@grep -h -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'
```
Comment (jzzheng22) [inline Makefile:32]:
```suggestion
	@grep -h -E '^[a-zA-Z_-]+:.*?## .*$$' $(MAKEFILE_LIST) | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m%-30s\033[0m %s\n", $$1, $$2}'

```

File: main.go [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		return
+	}
+	defer f.Close()
+	healthInfo := health.NewHealthInfo(&parameters)
+
+	// TODO: agentParameters - struct
+	simEnv := simulation.NewSimEnv(&parameters, healthInfo)
+	simEnv.Simulate()
+}
```
Comment (jzzheng22) [inline main.go:64]: Should we return an `err` here? So we know if the simulation failed?

**Review by jzzheng22 — CHANGES_REQUESTED** [review]

File: Makefile (hunk 9 above)
Comment (JaafarRammal) [reply Makefile:29]: This is pulling flags from the command line when you do `make run {some flags here}` and then appending on the `go run main.go {the flags here}`

Comment (JaafarRammal) [reply Makefile:29]: This is pulling flags from the command line when you do `make run {some flags here}` and then appending on the `go run main.go {the flags here}`

File: main.go (hunk 11 above)
Comment (JaafarRammal) [reply main.go:64]: Hmm yes I'll add a print I guess?

File: Makefile (hunk 9 above)
Comment (JaafarRammal) [reply Makefile:29]: for example if you want to run `go run main.go --config=/path/to/config` you would run `make run ARGS="--config=/path/to/config`

**Review by jzzheng22 — APPROVED** [review]

**Comment by udai-arneja** [comment]
"Never let your head hang down. Never give up and sit down and grieve. Find another way. And don’t pray when it rains if you don’t pray when the sun shines."

**Pull request merged**